### PR TITLE
QA-61 | Card Payment UI Tests

### PR DESCRIPTION
**Pull request by eugenezhernakov**
* Populate with similar UI tests to Android
* Add environment variables to test scheme
* Move strings to separate file
* Include CI/CD setup with Fastlane for signing & building for testing on Firebase Test Lab

### CODE REVIEW

File: Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/CardPayment/CardPaymentUITests.swift [hunk 1]
```diff
@@ -16,26 +16,347 @@ final class CardPaymentUITests: XCTestCase {
...
+        app.launch()
+        
+        app.cellWithIdentifier(TestData.PAY_WITH_CARD_LABEL)?.tap()
+        
+        app.fillCardSheetDetails(cardNumber: TestData.CARD_NUMBER,
+                             cardHolder: "Frictionless AuthFailed",
+                             expiryDate: TestData.CARD_EXPIRY,
+                             securityCode: TestData.CARD_SECURITY_CODE)
+        
+        app.cardDetailsSubmitButton?.tap()
```
Comment (eugenezhernakov) [inline Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/CardPayment/CardPaymentUITests.swift:267]: Same as [above](https://github.com/Judopay/JudoKit-iOS/pull/197/files#r1265346214)

File: fastlane/Matchfile [hunk 2]
```diff
@@ -0,0 +1,13 @@
+git_url("git@github.com:Judopay/JudoKit-iOS-SigningCertificates.git")
+
+storage_mode("git")
+
+type("adhoc") # The default type, can be: appstore, adhoc, enterprise or development
+
+app_identifier("com.judo.JudoPayDemoObjC")
+username("eugene.zhernakov@judopayments.com") # Your Apple Developer Portal username
```
Comment (stefan-tudor) [inline fastlane/Matchfile:8]: worth injecting this as an env variable instead

Comment (eugenezhernakov) [reply fastlane/Matchfile:8]: Done

File: Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/CardPayment/CardPaymentUITests.swift [hunk 3]
```diff
@@ -16,26 +16,347 @@ final class CardPaymentUITests: XCTestCase {
...
+        
+        tapCompleteButton(app)
+        
+        assertResultObject(app, "PreAuth", "AuthCode: ", "Success")
+    }
+}
+
+func assertResultObject(_ app: XCUIApplication, _ type: String, _ message: String, _ result: String) {
```
[the remark this answers is not in the document]
Comment (eugenezhernakov) [reply Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/CardPayment/CardPaymentUITests.swift]: Moved to separate file

File: Examples/ObjectiveCExampleApp/fastlane/.env [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+SCAN_CLEAN="true"
+SCAN_OUTPUT_DIRECTORY=$REPORTS_DIRECTORY
+SCAN_CODE_COVERAGE="true"
+
+# slather
+FL_SLATHER_PROJ="$CUSTOM_XCODE_PROJECT"
+FL_SLATHER_SCHEME="$GYM_SCHEME"
+FL_SLATHER_SOURCE_DIRECTORY="./Source"
+
+# oclint
```
Comment (RafalOzogJudopay) [inline Examples/ObjectiveCExampleApp/fastlane/.env:20]: Just cosmetic suggestion: maybe we don't need these "# gym", "# scan", "# slather", "# oclint" commented out titles in 4, 7, 14 and 20 lines - as it is repeated at the beginning of these constants anyway.

File: Examples/ObjectiveCExampleApp/fastlane/Matchfile [hunk 5]
```diff
@@ -0,0 +1,13 @@
+git_url("git@github.com:Judopay/JudoKit-iOS-SigningCertificates.git")
+
+storage_mode("git")
+
+type("development") # The default type, can be: appstore, adhoc, enterprise or development
+
+app_identifier(["com.judo.JudoPayDemoObjC", "com.judo.JudoPayDemoObjC.ObjectiveCExampleAppUITests.*"])
+username(ENV["APPLE_DEV_ACCOUNT"]) # Your Apple Developer Portal username
```
Comment (RafalOzogJudopay) [inline Examples/ObjectiveCExampleApp/fastlane/Matchfile:8]: Cosmetic, but probably we can remove this "# Your Apple Developer Portal username"

File: Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/Additions/Helpers.swift [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    XCTAssertEqual(resultValue, result, "Result value on result object does not match the expected string")
+}
+
+func tapCompleteButton(_ app: XCUIApplication) {
+    let completeButton = app.buttons["COMPLETE"]
+    XCTAssert(completeButton.waitForExistence(timeout: 10))
+        
+    var retryCount = 0
+    
+    while completeButton.exists && retryCount < 2 {
```
Comment (RafalOzogJudopay) [inline Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/Additions/Helpers.swift:41]: Is that correct that we expect two clicks in this case?

Comment (eugenezhernakov) [reply Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/Additions/Helpers.swift:41]: Unfortunately yes due to Global Pay sandbox environment, one click just refreshes the page.

Comment (RafalOzogJudopay) [reply Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/Additions/Helpers.swift:41]: That's fine. Maybe we could add some 1-liner comment about it in the code, if you'd like.

File: fastlane/Matchfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (eugenezhernakov) [reply fastlane/Matchfile:8]: I believe Stefan is making changes in this file as part of `improvements/ci-with-fastlane` branch and it's already been updated, just needs to sync with this branch =)

File: Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/Additions/XCUIApplication+Additions.swift [hunk 7]
```diff
@@ -51,4 +55,84 @@ extension XCUIApplication {
...
+    var editCardsButton: XCUIElement? {
+        get {
+            return buttonWithLabel("EDIT")
+        }
+    }
+    
+    var deleteCardButton: XCUIElement? {
+        get {
+            return buttonWithLabel("Delete")
+        }
```
Comment (RafalOzogJudopay) [inline Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/Additions/XCUIApplication+Additions.swift:115]: Maybe we could extract these buttons labels into constants, if you have time to do so. If you're busy, let's leave as it is.

File: Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/Additions/Helpers.swift (hunk 6 above)
Comment (eugenezhernakov) [reply Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/Additions/Helpers.swift:41]: I've added a comment for this

File: Examples/ObjectiveCExampleApp/fastlane/.env (hunk 4 above)
Comment (eugenezhernakov) [reply Examples/ObjectiveCExampleApp/fastlane/.env:20]: @stefan-tudor do you use these env vars?

File: Examples/ObjectiveCExampleApp/fastlane/Matchfile (hunk 5 above)
Comment (eugenezhernakov) [reply Examples/ObjectiveCExampleApp/fastlane/Matchfile:8]: Done

**Review by RafalOzogJudopay — CHANGES_REQUESTED** [review]
Overall, good code @eugenezhernakov 

Few brief points:

1. I left few comments about cosmetics (I leave it up to you to decide whether you would like to implement these changes or not), and 1 comment I would like you to have a look before I approve the PR (it's the comment about env variable in Matchfile). Please have a look.

2. I see this branch has some Fastlane setup files. Please ensure it won't collide with another Fastlane PR opened by Stefan this week (https://github.com/Judopay/JudoKit-iOS/pull/204).

3. I've seen your comment "Tried to assert that the snack bar is displayed with the expected message, but for some reason this doesn't work as it does above". Please do not hesitate to let me know if you would still need help with this, I am happy to help.

File: Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/Additions/XCUIApplication+Additions.swift (hunk 7 above)
Comment (eugenezhernakov) [reply Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/Additions/XCUIApplication+Additions.swift:131]: Done

File: Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/CardPayment/CardPaymentUITests.swift [hunk 8]
```diff
@@ -16,26 +16,339 @@ final class CardPaymentUITests: XCTestCase {
...
+        app.cardDetailsSubmitButton?.tap()
+        
+        let newCard = app.staticTexts["Visa Ending 7408 "]
+        XCTAssert(newCard.waitForExistence(timeout: 5), "Unable to add a new card")
+        
+        newCard.swipeLeft()
+        
+        app.deleteCardButton?.tap()
+        
+        app.deleteCardButton?.tap()
```
[the remark this answers is not in the document]
Comment (eugenezhernakov) [reply Examples/ObjectiveCExampleApp/ObjectiveCExampleAppUITests/CardPayment/CardPaymentUITests.swift:351]: Added

**Review by stefan-tudor — APPROVED** [review]

File: Examples/ObjectiveCExampleApp/fastlane/.env (hunk 4 above)
Comment (stefan-tudor) [reply Examples/ObjectiveCExampleApp/fastlane/.env]: Yes, I cleaned it up a bit, some was unused

**Review by RafalOzogJudopay — APPROVED** [review]

**Pull request merged**